pkg/kates/client.go

```diff
@@ -357,26 +358,71 @@ func isExpiredError(err error) bool {
 }
 
 type lw struct {
+	// All these fields are read-only and initialized on construction.
 	ctx           context.Context
 	client        dynamic.ResourceInterface
 	fieldSelector string
 	selector      string
-	synced        func()
+	synced        func(*lw)
 	once          sync.Once
+
+	// The mutex protects all the read-write fields.
+	mutex            sync.Mutex
+	initialListDone  bool
+	initialListCount int
+	addEventCount    int
 }
 
-func newListWatcher(ctx context.Context, client dynamic.ResourceInterface, fieldSelector, selector string, synced func()) cache.ListerWatcher {
+func newListWatcher(ctx context.Context, client dynamic.ResourceInterface, fieldSelector, selector string, synced func(*lw)) *lw {
 	return &lw{ctx: ctx, client: client, fieldSelector: fieldSelector, selector: selector, synced: synced}
 }
 
+func (lw *lw) withMutex(f func()) {
+	lw.mutex.Lock()
+	defer lw.mutex.Unlock()
+	f()
+}
+
+func (lw *lw) countAddEvent() {
+	lw.withMutex(func() {
+		lw.addEventCount++
+	})
+}
+
+// This computes whether we have synced a given watch. We used to use SharedInformer.HasSynced() for
+// this, but that seems to be a blatant lie that always return true. My best guess as to why it lies
+// is that it is actually reporting the synced state of an internal queue, but because the
+// SharedInformer mechanism adds another layer of dispatch on top of that internal queue, the
+// syncedness of that internal queue is irrelevant to whether enough layered events have been
+// dispatched to consider things synced at the dispatch layer.
+//
+// So to track syncedness properly for our users, when we do our first List() we remember how many
+// resourcees there are and we do not consider ourselves synced until we have dispatched at least as
+// many Add events as there are resources.
```

Comment:
Is this deterministic, or is there a way that it could conceivably still fail? (I _think_ it sounds deterministic, but e.g. suppose a given resource is deleted right as we're coming up -- are we certain to get an add and then a delete, or might we just never see anything for that resource?)

Reply:
I believe it is deterministic, because client-go computes the Add events based on the result of the List() method. It's client-go of course so my understanding could prove to be wrong, but that's the theory. ;-)

Reply:
Sounds sane, anyway. 🙂